src/core/symbology-ng/qgsfillsymbollayerv2.cpp

```diff
@@ -3432,9 +3436,38 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
 {
   Q_UNUSED( rings );
 
-  if ( !mPointOnAllParts &&
-       context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_NUM ).toInt() != context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_BIGGEST ).toInt() )
-    return;
+  if ( !mPointOnAllParts )
+  {
+    const QgsFeature* feature = context.feature();
+    if ( feature )
+    {
+      if ( feature->id() != mBiggestPartFeatureId )
+      {
+        mBiggestPartFeatureId = feature->id();
+        mBiggestPartIndex = 1;
+
+        unsigned int part_count = context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_COUNT ).toUInt();
+        if ( part_count > 1 )
+        {
+          const QgsGeometry *geom = feature->constGeometry();
+          const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+          double area_biggest = 0;
+          for ( unsigned int i = 0; i < part_count; ++i )
+          {
+            if ( geomCollection->geometryN( i )->area() > area_biggest )
+            {
+              area_biggest = geomCollection->geometryN( i )->area();
+              mBiggestPartIndex = i + 1;
+            }
+          }
+        }
+      }
+
+      if ( context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_NUM ).toInt() != mBiggestPartIndex )
```

Comment:
This is already available in the expression context and not new. Is your idea to pass them twice, once in the expression context and once in the render context? Just have to make sure that this also works below geometry generators which can work cleanly with expression context (just push and pop scopes) whereas with the rendering context this cannot just be popped again.